app/Listeners/TaskUpdateXP.php

```diff
@@ -26,66 +26,64 @@ public function __construct()
+        $this->model = $event->model;
 
         //task user id
-        $user_id = $this->tasks->task_history[0]['user'];
+        $userId = $this->model->task_history[0]['user'];
 
         //project owner id
-        $project_owner_id = $this->tasks->owner;
+        $projectOwnerId = $this->model->ownerId;
 
-        //get task's XP value
-        $task_xp = $this->tasks->xp;
+        $userProfile = Profile::find($userId);
+        $ownerProfile = Profile::find($projectOwnerId);
 
-        //
-        $user_profile = Profile::find($user_id);
-        var_dump($user_profile);
-        $owner_profile = Profile::find($project_owner_id);
 
-        $end = 0;
+        //get task's XP value
+        $taskXp = $this->model->xp;
+
+        $work = 0;
         $review = 0;
-        for ($i = 0; $i < count($this->tasks->task_history); $i++) {
-            if ($this->tasks->submitted_for_qa === true) {
-                $end += $this->tasks->task_history[$i]['timestamp'] - $this->tasks->task_history[$i - 1]['timestamp'];
-                echo $end;
-            } elseif ($this->tasks->submitted_for_qa === false) {
-                $review += $this->tasks->task_history[$i]['timestamp'] - $this->tasks->task_history[$i - 1]['timestamp'];
+        for ($i = count($this->model->task_history) - 1; $i >= 0; $i--) {
+            if (((($this->model->task_history[$i]['event'] == "Task returned for development")) || $this->model->task_history[$i]['event'] == "Task passed QA!") && ($i > 0)) {
+                    $review += floor($this->model->task_history[$i]['timestamp'] / 1000) - floor($this->model->task_history[$i - 1]['timestamp'] / 1000);
+            } elseif (($this->model->task_history[$i]['event'] == "Task submitted for QA") && ($i > 0)) {
+                    $work += floor($this->model->task_history[$i]['timestamp'] / 1000) - floor($this->model->task_history[$i - 1]['timestamp'] / 1000);
             }
         }
 
+        //if time spent reviewing code more than 1 day, deduct project/task owner 3 XP
+        if ($review > 24 * 60 * 60) {
+            $ownerProfile->xp -= 3;
+            $ownerProfile->save();
+        }
 
-
-
-        $coefficient = ($end / $this->tasks->estimatedHours);
-
-        echo $coefficient;
+        //apply xp change
+        $coefficient = number_format(($work / ($this->model->estimatedHours * 60 * 60)), 5);
         switch ($coefficient) {
             case ($coefficient <= 0.75):
-                $user_profile->xp += $task_xp + 3;
-                $user_profile->save();
+                $userProfile->xp += $taskXp + 3;
+                $userProfile->save();
                 break;
             case ($coefficient >= 0.75 && $coefficient <= 1):
-                $user_profile->xp += $task_xp;
-                $user_profile->save();
+                $userProfile->xp += $taskXp;
+                $userProfile->save();
                 break;
             case ($coefficient >= 1.01 && $coefficient <= 1.1):
```

Comment:
should be `$coefficient > 1 && $coefficient <= 1` in case `$coefficient = 1.001` for example - that wouldn't be accounted for. Same goes for rest of the range checks here